std-bits/table/src/main/java/org/enso/table/operations/AddRowNumber.java

```diff
@@ -1,144 +1,96 @@
-    if (groupingColumns.length == 0) {
-      throw new IllegalArgumentException("At least one grouping column is required.");
-    }
-
-    int n = groupingColumns[0].getSize();
-    long[] numbers = new long[n];
-    Storage<?>[] groupingStorages =
-        Arrays.stream(groupingColumns).map(Column::getStorage).toArray(Storage[]::new);
-    ColumnAggregatedProblemAggregator groupingProblemAggregator =
-        new ColumnAggregatedProblemAggregator(problemAggregator);
-    List<TextFoldingStrategy> textFoldingStrategy =
-        ConstantList.make(TextFoldingStrategy.unicodeNormalizedFold, groupingStorages.length);
-    Map<UnorderedMultiValueKey, RangeIterator> groups = new HashMap<>();
-    for (int i = 0; i < n; i++) {
-      UnorderedMultiValueKey key =
-          new UnorderedMultiValueKey(groupingStorages, i, textFoldingStrategy);
-      key.checkAndReportFloatingEquality(
-          groupingProblemAggregator, columnIx -> groupingColumns[columnIx].getName());
-      RangeIterator it = groups.computeIfAbsent(key, k -> new RangeIterator(start, step));
-      numbers[i] = it.next();
-    }
-    return new LongStorage(numbers, IntegerType.INT_64);
-  }
-
-  public static LongStorage create_ordered_numbering(
-      long start, long step, Column[] orderingColumns, int[] directions) {
-    if (orderingColumns.length == 0) {
-      throw new IllegalArgumentException("At least one ordering column is required.");
-    }
-    if (orderingColumns.length != directions.length) {
-      throw new IllegalArgumentException(
-          "The number of ordering columns and directions must be the same.");
-    }
-
-    int n = orderingColumns[0].getSize();
-    Storage<?>[] orderingStorages =
-        Arrays.stream(orderingColumns).map(Column::getStorage).toArray(Storage[]::new);
-    long[] numbers = new long[n];
-    List<OrderedMultiValueKey> keys =
-        new ArrayList<>(
-            IntStream.range(0, n)
-                .mapToObj(i -> new OrderedMultiValueKey(orderingStorages, i, directions))
-                .toList());
-
-    keys.sort(null);
-
-    RangeIterator it = new RangeIterator(start, step);
-    for (var key : keys) {
-      numbers[key.getRowIndex()] = it.next();
-    }
-    return new LongStorage(numbers, IntegerType.INT_64);
-  }
-
-  public static LongStorage create_grouped_ordered_numbering(
+  public static Storage<?> create_numbering(
       long start,
       long step,
+      Column[] groupingColumns,
       Column[] orderingColumns,
       int[] directions,
-      Column[] groupingColumns,
       ProblemAggregator problemAggregator) {
-    if (orderingColumns.length == 0) {
-      throw new IllegalArgumentException("At least one ordering column is required.");
-    }
     if (orderingColumns.length != directions.length) {
       throw new IllegalArgumentException(
           "The number of ordering columns and directions must be the same.");
     }
```

Comment:
Shall we move this check to be directly in `RunningLooper`? Then we will have this check also in one place.

Reply:
Done